### ISSUE TITLE
feat(agent): extend DagobertDuckDQN with cumulative Task 2 capabilities

### ISSUE THREAD

**Issue by 1BlauNitrox** [1BlauNitrox/mle-final-project#44]
## Problem statement

`DagobertDuckDQN` currently implements only Task 1 visible-coin navigation with
five actions and no bomb, crate, explosion, or survival representation. Task 2
requires one cumulative learned policy that retains Task 1 navigation while
learning to destroy crates, reveal and collect hidden coins, place bombs, and
avoid killing itself on a board without opponents.

Replacing the frozen Task 1 policy or training an unrelated agent from scratch
would remove the lineage needed for regression evidence. The Task 2 policy must
therefore be implemented in the self-contained successor prepared by #43 and
must explicitly migrate from the frozen Task 1 baseline in #42.

## Proposed solution

Extend `agent_code/DagobertDuckDQNTask2/` with a versioned Task 2 feature schema,
six-action DQN output, controlled checkpoint migration, Task 2 rewards, training
diagnostics, and tests. Curriculum orchestration is tracked separately in #50
so that the agent-side capability can be reviewed independently from shared
Tasks 2-4 tooling.

Retain the Task 1 navigation features as a documented prefix. Add at minimum:

- bomb availability;
- current and near-future blast danger derived from bomb timers, walls, crates,
  and the explosion map;
- safe movement directions and reachable escape information;
- whether a safe escape exists after a hypothetical bomb placement;
- direction and reachable distance to useful crates;
- number of crates affected by a bomb at the current position;
- whether a bomb has a useful crate target.

The action order is fixed for this successor as:

```text
UP, RIGHT, DOWN, LEFT, WAIT, BOMB
```

Appending `BOMB` preserves the five Task 1 action indices. Any action mask must
be deterministic, documented, tested, and limited to impossible or explicitly
defined safety/usefulness conditions; it must not implement the complete policy
instead of the learned model.

Task 1 checkpoint migration must be explicit and one-way:

- copy matching Task 1 input-layer weights into the corresponding Task 2 input
  columns;
- initialize new Task 2 input columns using a documented seeded rule;
- preserve compatible hidden-layer weights and biases;
- copy the five Task 1 output rows without reordering;
- initialize the `BOMB` output row using a documented conservative rule;
- initialize the target network from the migrated online network;
- reset incompatible optimizer and replay state rather than coercing shapes;
- record the parent artifact SHA-256 and migration configuration.

Centralize implementation-default rewards for `COIN_COLLECTED`,
`CRATE_DESTROYED`, `COIN_FOUND`, `INVALID_ACTION`, `WAITED`, `KILLED_SELF`,
`GOT_KILLED`, and `SURVIVED_ROUND`. Custom shaping must be derived from state
transitions, documented, and free of evaluation-time dependencies. Reward
values are implementation defaults only until tested in a prospective
experiment.

The training-only curriculum that mixes `coin-heaven`, an easier crate scenario
such as `loot-crate`, and `classic` with no opponents is explicitly delegated to
#50. This issue supplies the self-contained agent capability and diagnostics
that the shared orchestration will drive; evaluation remains entirely
self-contained.

## Acceptance criteria

- [ ] An owner and non-author reviewer are assigned.
- [ ] #42 is frozen and #43 has produced the self-contained Task 2 successor
      before implementation starts.
- [ ] Task 1 source and artifact remain unchanged; the successor records their
      commit and SHA-256 as parent lineage.
- [ ] The feature schema has an explicit version, stable order, normalization,
      expected shape, value domains, and neutral conventions for absent crates
      or bombs.
- [ ] The existing Task 1 navigation features remain present and are tested for
      equivalent values on Task 1 states.
- [ ] Blast propagation respects stone walls, crate pass-through, bomb timers,
      overlapping danger windows, and the framework explosion map.
- [ ] Escape features use reachable paths and distinguish currently safe tiles
      from tiles that become unsafe before arrival.
- [ ] The model emits exactly six Q-values in the documented action order and
      never returns a string outside that set.
- [ ] A schema-validated migration copies compatible Task 1 weights and action
      outputs, initializes new parameters deterministically, resets incompatible
      optimizer/replay state, and records the parent checksum.
- [ ] Migration tests verify preserved old input/output parameters, initialized
      new parameters, target synchronization, deterministic repeatability, and
      rejection of incompatible parents.
- [ ] Unit tests cover danger-map edge cases, blast blocking, timers, safe
      escape, no-escape states, crate targeting, bomb availability, neutral
      Task 1 states, and feature tensor shape/range.
- [ ] Training tests cover ordinary and terminal Task 2 transitions, bomb
      actions, reward calculation, target masking, replay insertion, optimizer
      updates, resume behavior, and bounded replay.
- [ ] Evaluation disables exploration, gradients, replay insertion, optimizer
      updates, target updates, and file writes.
- [ ] Persistence remains relative, atomic, restricted, schema-validated, and
      sufficient for deterministic evaluation and deliberate training resume.
- [ ] Training diagnostics expose coins collected/found, crates destroyed,
      bombs dropped, useful bombs where defined, self-kills, survival, invalid
      actions, loss/TD error, epsilon, replay size, and target updates.
- [ ] All implementation-default hyperparameters/rewards are centralized and
      documented before smoke training. Curriculum orchestration remains an
      explicit downstream deliverable of #50.
- [ ] A seeded headless Task 2 training smoke creates a compatible non-empty
      artifact and expected metrics.
- [ ] A subsequent read-only evaluation smoke on `classic` without opponents
      completes deterministically and preserves artifact bytes.
- [ ] A Task 1 `coin-heaven` regression smoke completes and selects `BOMB` zero
      times; this is integration evidence, not a non-inferiority claim.
- [ ] A clean-upstream package with only the successor and declared
      requirements runs without modified framework files.
- [ ] CPU-only evaluation uses one thread, no multiprocessing, less than 8 GB
      RAM, p95 decision time below `50 ms`, and maximum below `100 ms`.
- [ ] Agent card, architecture documentation, dependencies, migration lineage,
      limitations, and AI disclosure are updated.
- [ ] No smoke result is presented as Task 2 performance evidence.
- [ ] Ruff, compilation, pytest, packaging, smoke, and GitHub Actions checks
      pass.

## Tournament or framework constraints

- Evaluation must run in an unchanged upstream framework using only files and
  declared dependencies inside the packaged agent.
- The final policy is CPU-only, single-threaded, non-multiprocessing, below
  8 GB RAM, and comfortably below the 0.5-second action limit.
- Training may use additional compute, but no training-only dependency or state
  may be required at evaluation.
- Paths remain relative to the agent module; model loading must fail closed on
  incompatible or corrupt artifacts.
- The policy must remain genuinely learned. Feature engineering and safety
  masks may support learning but must not become a complete rule-based agent.

## Dependencies or blockers

Blocked by:

- #42, frozen Task 1 DQN source and artifact;
- #43, cumulative self-contained successor architecture;
- the parent artifact and migration manifest being available;
- agreement on the exact feature order and reward defaults before coding.

Downstream:

- #50 will provide the shared Task 2 curriculum orchestration;
- the Task 2 DQN experiment will evaluate new capability and paired Task 1
  retention;
- Task 3 work must extend this successor rather than the frozen Task 1 agent.

Required resources:

- normal development/CI resources plus short seeded smoke runs;
- no long-running scientific training is authorized by this implementation
  issue.

Out of scope:

- opponents, hunting, self-play, and Task 3/4 behavior;
- scientific hyperparameter/reward/feature optimization;
- final held-out evaluation or Task 2 completion claims;
- modifying the frozen Task 1 baseline.

**Comment by 1BlauNitrox** [comment]
Shared curriculum and experiment-matrix orchestration is now tracked in #50. The downstream DQN experiment against coin_collector_agent is #51 and remains blocked by this Task 2 implementation plus #46.
